### DISCUSSION TITLE
Is it possible to set parameters (or functions) with the same absolute value but different signs depending on the state?

### DISCUSSION THREAD

**Discussion by YusukeGoto510** [bmcclintock/momentuHMM#146] [Q&A] [answered]
I’m trying to classify a time series of a variable x with a covariant y into 3 states. I assume that the mean value of x in states 1 and 2 is a spline function of y, and specify DM as follows:

DM<-list(x= list(mean = ~state1(bs(y, df=5, degree=3)) + state2(bs(y, df=5,degree=3)), sd = ~1)).

However, I found that the behavior of x is symmetric around x = 0, 
so if the mean of state 1 is f(y), 
then I want to fix the mean of state 2 to be exactly the same as state 1, but differs only in sign, i.e. -f(y). 

I would appreciate it if you could show me how this can be implemented.
Thank you in advance for your time!

**Comment by bmcclintock** [comment, marked as answer]
I'm not sure I totally understand what you're trying to do, but means for a normal distribution can be forced to be "opposite" by imposing constraints on the working-scale mean parameters via the `DM` argument. I'll show how this can be done for demonstration purposes, but please note my word of caution about HMMs with "opposite" symmetric state-dependent distributions that are centered at zero and have the same variance.
```
library(momentuHMM)

T <- 10000
N <- 1

covs <- data.frame(cov1 = rnorm(T*N))

DM <- list(x=matrix(c( 1,    "cov1",0,0,
                      -1,"I(-cov1)",0,0,
                       0,         0,1,0,
                       0,         0,0,1),4,4,byrow=TRUE,
                     dimnames=list(c("mean_1","mean_2","sd_1","sd_2"),
                                   c("mean:(Intercept)","mean:cov1","sd_1:(Intercept)","sd_2:(Intercept)"))))

Par0 <- list(x=c(0,0.5,log(1),log(1)))

beta0 <- matrix(c(-1.5,-1.5),ncol=2,byrow=TRUE)

sim <- simData(nbAnimals = N,obsPerAnimal = T,dist=list(x="norm"),DM=DM,Par=Par0,beta=beta0,covs=covs,states=TRUE)

summary(sim[which(sim$states==1),"x"])
summary(sim[which(sim$states==2),"x"])

fit <- fitHMM(sim,nbStates=2,dist=list(x="norm"),DM=DM,Par0=Par0,nlmPar=list(print.level=2))
```
B-splines make it a bit more complicated. There might be an exact way to specify the "mean" of the smooth to be zero, but this hack is pretty close:
```
library(splines)

covs[,paste0("bs",1:3)] <- bs(covs$cov1)

DM <- list(x=matrix(c(1,    "bs1",    "bs2",    "bs3",0,0,
                     -1,"I(-bs1)","I(-bs2)","I(-bs3)",0,0,
                      0,        0,        0,        0,1,0,
                      0,        0,        0,        0,0,1),4,6,byrow=TRUE,
                     dimnames=list(c("mean_1","mean_2","sd_1","sd_2"),
                                   c("mean:(Intercept)",paste0("mean:bs",1:3),"sd_1:(Intercept)","sd_2:(Intercept)"))))

Par0 <- list(x=c(NA,0.5,1,2,log(1),log(1)))
Par0$x[1] <- -sum(apply(covs,2,mean)[2:4]*Par0$x[2:4])

beta0 <- matrix(c(-1.5,-1.5),ncol=2,byrow=TRUE)

sim <- simData(nbAnimals = N,obsPerAnimal = T,dist=list(x="norm"),DM=DM,Par=Par0,beta=beta0,covs=covs,states=TRUE)

summary(sim[which(sim$states==1),"x"])
summary(sim[which(sim$states==2),"x"])

fit <- fitHMM(sim,nbStates=2,dist=list(x="norm"),DM=DM,Par0=Par0,nlmPar=list(print.level=2))
```
At least one word of caution though -- for "opposite" symmetric distributions centered at zero with the same `sd` (such that the only difference between them is that the observed values are opposite in sign as in the example above), the fitted model is unlikely to properly identify these states (i.e. they will be unidentifiable). You can verify this by comparing the Viterbi-decoded states (or estimated state probabilities) to the "true" simulated states:
```
mean(sim$states==viterbi(fit))
mean(apply(stateProbs(fit),1,which.max)==sim$states)
```
